{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Notebook env: seq_trblsht (seq_trblsht kernel)\n",
    "## This notebook is used to analyze the fraction of reads that map to the guide region in GEM1 of the CRISPRa library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRa_repeat_screen/low_guides_trblsht/STAR_guide_reads_alignment\n",
    "nohup STAR --outSAMtype BAM SortedByCoordinate --readFilesCommand zcat --runThreadN 16 --genomeDir /home/genomes/hg38/star_hg38 --readFilesIn /home/ssobti/projects/heterogeneity_brian/data/072623_MD231_CRISPRa_repeat_screen/BWHG37/SL424_S3_L004_R2_001.fastq.gz --outFileNamePrefix SL424_R2_all_reads_no_trim &"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## not run\n",
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRa_repeat_screen/low_guides_trblsht/STAR_guide_reads_alignment\n",
    "samtools index --threads 16 SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bam SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bai"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## not run\n",
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRa_repeat_screen/low_guides_trblsht/STAR_guide_reads_alignment\n",
    "nohup bedtools multicov -bams SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bam -bed /home/genomes/hg38/star_hg38/Homo_sapiens.GRCh38.99.gtf > SL424_R2_all_reads_no_trim_bedtools.txt &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRa_repeat_screen/low_guides_trblsht/STAR_guide_reads_alignment\n",
    "nohup htseq-count -n 16 -f bam -s yes -i gene_id SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bam /home/genomes/hg38/star_hg38/Homo_sapiens.GRCh38.99.gtf > SL424_R2_all_reads_no_trim_HTSeq_count.txt &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60680</th>\n",
       "      <td>__alignment_not_unique</td>\n",
       "      <td>84789687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60676</th>\n",
       "      <td>__no_feature</td>\n",
       "      <td>71026957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60677</th>\n",
       "      <td>__ambiguous</td>\n",
       "      <td>5641588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21453</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>2100940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21417</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>1854876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28556</th>\n",
       "      <td>ENSG00000229190</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28557</th>\n",
       "      <td>ENSG00000229191</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28558</th>\n",
       "      <td>ENSG00000229192</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28559</th>\n",
       "      <td>ENSG00000229195</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30340</th>\n",
       "      <td>ENSG00000231794</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60681 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gene     count\n",
       "60680  __alignment_not_unique  84789687\n",
       "60676            __no_feature  71026957\n",
       "60677             __ambiguous   5641588\n",
       "21453         ENSG00000211459   2100940\n",
       "21417         ENSG00000210082   1854876\n",
       "...                       ...       ...\n",
       "28556         ENSG00000229190         0\n",
       "28557         ENSG00000229191         0\n",
       "28558         ENSG00000229192         0\n",
       "28559         ENSG00000229195         0\n",
       "30340         ENSG00000231794         0\n",
       "\n",
       "[60681 rows x 2 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## load in bed file gene counts from reads\n",
    "import pandas as pd\n",
    "import mygene as mg\n",
    "\n",
    "path = '/home/ssobti/projects/heterogeneity_brian/output_data/CRISPRa_repeat_screen/low_guides_trblsht/STAR_guide_reads_alignment/'\n",
    "read_counts = pd.read_csv(path + 'SL424_R2_all_reads_no_trim_HTSeq_count.txt', sep='\\t', names = ['gene', 'count'])\n",
    "read_counts.sort_values(by='count', ascending=False, inplace=True)\n",
    "read_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21453</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>2100940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21417</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>1854876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1315</td>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>398501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6598</td>\n",
       "      <td>ENSG00000132507</td>\n",
       "      <td>359221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17941</td>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>324730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60673</th>\n",
       "      <td>28556</td>\n",
       "      <td>ENSG00000229190</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60674</th>\n",
       "      <td>28557</td>\n",
       "      <td>ENSG00000229191</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60675</th>\n",
       "      <td>28558</td>\n",
       "      <td>ENSG00000229192</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60676</th>\n",
       "      <td>28559</td>\n",
       "      <td>ENSG00000229195</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60677</th>\n",
       "      <td>30340</td>\n",
       "      <td>ENSG00000231794</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60678 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index             gene    count\n",
       "0      21453  ENSG00000211459  2100940\n",
       "1      21417  ENSG00000210082  1854876\n",
       "2       1315  ENSG00000075624   398501\n",
       "3       6598  ENSG00000132507   359221\n",
       "4      17941  ENSG00000198938   324730\n",
       "...      ...              ...      ...\n",
       "60673  28556  ENSG00000229190        0\n",
       "60674  28557  ENSG00000229191        0\n",
       "60675  28558  ENSG00000229192        0\n",
       "60676  28559  ENSG00000229195        0\n",
       "60677  30340  ENSG00000231794        0\n",
       "\n",
       "[60678 rows x 3 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove_rows = ['__alignment_not_unique', '__no_feature', '__ambiguous']\n",
    "read_counts = read_counts[~read_counts['gene'].isin(remove_rows)].reset_index()\n",
    "read_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "querying 2001-3000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 7001-8000...done.\n",
      "querying 8001-9000...done.\n",
      "querying 9001-10000...done.\n",
      "querying 10001-11000...done.\n",
      "querying 11001-12000...done.\n",
      "querying 12001-13000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 14001-15000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 19001-20000...done.\n",
      "querying 20001-21000...done.\n",
      "querying 21001-22000...done.\n",
      "querying 22001-23000...done.\n",
      "querying 23001-24000...done.\n",
      "querying 24001-25000...done.\n",
      "querying 25001-26000...done.\n",
      "querying 26001-27000...done.\n",
      "querying 27001-28000...done.\n",
      "querying 28001-29000...done.\n",
      "querying 29001-30000...done.\n",
      "querying 30001-31000...done.\n",
      "querying 31001-32000...done.\n",
      "querying 32001-33000...done.\n",
      "querying 33001-34000...done.\n",
      "querying 34001-35000...done.\n",
      "querying 35001-36000...done.\n",
      "querying 36001-37000...done.\n",
      "querying 37001-38000...done.\n",
      "querying 38001-39000...done.\n",
      "querying 39001-40000...done.\n",
      "querying 40001-41000...done.\n",
      "querying 41001-42000...done.\n",
      "querying 42001-43000...done.\n",
      "querying 43001-44000...done.\n",
      "querying 44001-45000...done.\n",
      "querying 45001-46000...done.\n",
      "querying 46001-47000...done.\n",
      "querying 47001-48000...done.\n",
      "querying 48001-49000...done.\n",
      "querying 49001-50000...done.\n",
      "querying 50001-51000...done.\n",
      "querying 51001-52000...done.\n",
      "querying 52001-53000...done.\n",
      "querying 53001-54000...done.\n",
      "querying 54001-55000...done.\n",
      "querying 55001-56000...done.\n",
      "querying 56001-57000...done.\n",
      "querying 57001-58000...done.\n",
      "querying 58001-59000...done.\n",
      "querying 59001-60000...done.\n",
      "querying 60001-60678...done.\n",
      "Finished.\n",
      "267 input query terms found no hit:\n",
      "\t['ENSG00000254615', 'ENSG00000130723', 'ENSG00000239665', 'ENSG00000205485', 'ENSG00000112096', 'ENS\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"
     ]
    }
   ],
   "source": [
    "## convert ENSEMBL IDs to gene symbols\n",
    "import mygene as mg\n",
    "mg = mygene.MyGeneInfo()\n",
    "gene_symbols = mg.querymany(read_counts['gene'].tolist(), scopes='ensembl.gene', fields = ['symbol'], species = 'human')\n",
    "gene_symbols = [gene['symbol'] if 'symbol' in gene.keys() else 'NA' for gene in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21453</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>2100940</td>\n",
       "      <td>RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21417</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>1854876</td>\n",
       "      <td>RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1315</td>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>398501</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6598</td>\n",
       "      <td>ENSG00000132507</td>\n",
       "      <td>359221</td>\n",
       "      <td>EIF5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17941</td>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>324730</td>\n",
       "      <td>COX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7539</td>\n",
       "      <td>ENSG00000137818</td>\n",
       "      <td>315297</td>\n",
       "      <td>RPLP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>17805</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>276649</td>\n",
       "      <td>ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>17572</td>\n",
       "      <td>ENSG00000197956</td>\n",
       "      <td>265503</td>\n",
       "      <td>S100A6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>39796</td>\n",
       "      <td>ENSG00000251562</td>\n",
       "      <td>232729</td>\n",
       "      <td>MALAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9263</td>\n",
       "      <td>ENSG00000149925</td>\n",
       "      <td>231729</td>\n",
       "      <td>ALDOA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>14990</td>\n",
       "      <td>ENSG00000181163</td>\n",
       "      <td>201603</td>\n",
       "      <td>NPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21433</td>\n",
       "      <td>ENSG00000210184</td>\n",
       "      <td>195655</td>\n",
       "      <td>TRNS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3492</td>\n",
       "      <td>ENSG00000108107</td>\n",
       "      <td>166608</td>\n",
       "      <td>RPL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10443</td>\n",
       "      <td>ENSG00000160213</td>\n",
       "      <td>151362</td>\n",
       "      <td>CSTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>8290</td>\n",
       "      <td>ENSG00000142937</td>\n",
       "      <td>142812</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>37366</td>\n",
       "      <td>ENSG00000245532</td>\n",
       "      <td>141355</td>\n",
       "      <td>NEAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16238</td>\n",
       "      <td>ENSG00000186468</td>\n",
       "      <td>137159</td>\n",
       "      <td>RPS23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17848</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>134699</td>\n",
       "      <td>ND5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17911</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>128368</td>\n",
       "      <td>ND1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>17882</td>\n",
       "      <td>ENSG00000198840</td>\n",
       "      <td>125019</td>\n",
       "      <td>ND3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>452</td>\n",
       "      <td>ENSG00000026025</td>\n",
       "      <td>124809</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1718</td>\n",
       "      <td>ENSG00000087086</td>\n",
       "      <td>109843</td>\n",
       "      <td>FTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3949</td>\n",
       "      <td>ENSG00000111775</td>\n",
       "      <td>109196</td>\n",
       "      <td>COX6A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>22075</td>\n",
       "      <td>ENSG00000212907</td>\n",
       "      <td>108345</td>\n",
       "      <td>ND4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>5611</td>\n",
       "      <td>ENSG00000125148</td>\n",
       "      <td>106854</td>\n",
       "      <td>MT2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>17909</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>105855</td>\n",
       "      <td>ND4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>5805</td>\n",
       "      <td>ENSG00000126261</td>\n",
       "      <td>95559</td>\n",
       "      <td>UBA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>767</td>\n",
       "      <td>ENSG00000057019</td>\n",
       "      <td>86595</td>\n",
       "      <td>DCBLD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>14123</td>\n",
       "      <td>ENSG00000176731</td>\n",
       "      <td>86114</td>\n",
       "      <td>RBIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3913</td>\n",
       "      <td>ENSG00000111640</td>\n",
       "      <td>85755</td>\n",
       "      <td>GAPDH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index             gene    count  symbol\n",
       "0   21453  ENSG00000211459  2100940    RNR1\n",
       "1   21417  ENSG00000210082  1854876    RNR2\n",
       "2    1315  ENSG00000075624   398501    ACTB\n",
       "3    6598  ENSG00000132507   359221   EIF5A\n",
       "4   17941  ENSG00000198938   324730    COX3\n",
       "5    7539  ENSG00000137818   315297   RPLP1\n",
       "6   17805  ENSG00000198695   276649     ND6\n",
       "7   17572  ENSG00000197956   265503  S100A6\n",
       "8   39796  ENSG00000251562   232729  MALAT1\n",
       "9    9263  ENSG00000149925   231729   ALDOA\n",
       "10  14990  ENSG00000181163   201603    NPM1\n",
       "11  21433  ENSG00000210184   195655   TRNS2\n",
       "12   3492  ENSG00000108107   166608   RPL28\n",
       "13  10443  ENSG00000160213   151362    CSTB\n",
       "14   8290  ENSG00000142937   142812    RPS8\n",
       "15  37366  ENSG00000245532   141355   NEAT1\n",
       "16  16238  ENSG00000186468   137159   RPS23\n",
       "17  17848  ENSG00000198786   134699     ND5\n",
       "18  17911  ENSG00000198888   128368     ND1\n",
       "19  17882  ENSG00000198840   125019     ND3\n",
       "20    452  ENSG00000026025   124809     VIM\n",
       "21   1718  ENSG00000087086   109843     FTL\n",
       "22   3949  ENSG00000111775   109196  COX6A1\n",
       "23  22075  ENSG00000212907   108345    ND4L\n",
       "24   5611  ENSG00000125148   106854    MT2A\n",
       "25  17909  ENSG00000198886   105855     ND4\n",
       "26   5805  ENSG00000126261    95559    UBA2\n",
       "27    767  ENSG00000057019    86595  DCBLD2\n",
       "28  14123  ENSG00000176731    86114    RBIS\n",
       "29   3913  ENSG00000111640    85755   GAPDH"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## adding gene symbols\n",
    "read_counts['symbol'] = gene_symbols\n",
    "read_counts[0:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "seq_trblsht",
   "language": "python",
   "name": "seq_trblsht"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
