{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T16:34:24.953093Z",
     "iopub.status.busy": "2023-08-29T16:34:24.952737Z",
     "iopub.status.idle": "2023-08-29T16:34:24.958475Z",
     "shell.execute_reply": "2023-08-29T16:34:24.957352Z"
    }
   },
   "outputs": [],
   "source": [
    "## Notebook env: seq_trblsht (seq_trblsht kernel)\n",
    "## This notebook is used to analyze the fraction of reads that map to the guide region in GEM1 of the CRISPRa library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T16:34:24.962476Z",
     "iopub.status.busy": "2023-08-29T16:34:24.962166Z",
     "iopub.status.idle": "2023-08-29T16:41:31.385207Z",
     "shell.execute_reply": "2023-08-29T16:41:31.383895Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aug 29 09:34:25 ..... started STAR run\n",
      "Aug 29 09:34:25 ..... loading genome\n",
      "Aug 29 09:34:32 ..... started mapping\n",
      "Aug 29 09:41:10 ..... finished mapping\n",
      "Aug 29 09:41:11 ..... started sorting BAM\n",
      "Aug 29 09:41:30 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRi_CRISPRa_cell_culture_screen/CRISPRi_STAR_guide_reads_alignment\n",
    "nohup STAR --outSAMtype BAM SortedByCoordinate --readFilesCommand zcat --runThreadN 16 --genomeDir /home/genomes/hg38/star_hg38 --readFilesIn /home/ssobti/projects/heterogeneity_brian/data/032923_MD231_CRISPRi_CRISPRa_screen/230329_A01102_0587_AH3CY5DSX7_S4_200/BWHG34/399_S6_L001_R2_001.fastq.gz --outFileNamePrefix SL399_R2_all_reads_no_trim &"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## not run\n",
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRi_CRISPRa_cell_culture_screen/CRISPRi_STAR_guide_reads_alignment\n",
    "samtools index --threads 16 SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bam SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bai"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## not run\n",
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRi_CRISPRa_cell_culture_screen/CRISPRi_STAR_guide_reads_alignment\n",
    "nohup bedtools multicov -bams SL424_R2_all_reads_no_trimAligned.sortedByCoord.out.bam -bed /home/genomes/hg38/star_hg38/Homo_sapiens.GRCh38.99.gtf > SL424_R2_all_reads_no_trim_bedtools.txt &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T16:41:31.391692Z",
     "iopub.status.busy": "2023-08-29T16:41:31.391112Z",
     "iopub.status.idle": "2023-08-29T17:00:16.229977Z",
     "shell.execute_reply": "2023-08-29T17:00:16.228607Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for 'SL399_R2_all_reads_no_trimAligned.sortedByCoord.out.bam'\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "2905054 GFF lines processed.\n",
      "[E::idx_find_and_load] Could not retrieve index file for 'SL399_R2_all_reads_no_trimAligned.sortedByCoord.out.bam'\n",
      "100000 alignment records processed.\n",
      "200000 alignment records processed.\n",
      "300000 alignment records processed.\n",
      "400000 alignment records processed.\n",
      "500000 alignment records processed.\n",
      "600000 alignment records processed.\n",
      "700000 alignment records processed.\n",
      "800000 alignment records processed.\n",
      "900000 alignment records processed.\n",
      "1000000 alignment records processed.\n",
      "1100000 alignment records processed.\n",
      "1200000 alignment records processed.\n",
      "1300000 alignment records processed.\n",
      "1400000 alignment records processed.\n",
      "1500000 alignment records processed.\n",
      "1600000 alignment records processed.\n",
      "1700000 alignment records processed.\n",
      "1800000 alignment records processed.\n",
      "1900000 alignment records processed.\n",
      "2000000 alignment records processed.\n",
      "2100000 alignment records processed.\n",
      "2200000 alignment records processed.\n",
      "2300000 alignment records processed.\n",
      "2400000 alignment records processed.\n",
      "2500000 alignment records processed.\n",
      "2600000 alignment records processed.\n",
      "2700000 alignment records processed.\n",
      "2800000 alignment records processed.\n",
      "2900000 alignment records processed.\n",
      "3000000 alignment records processed.\n",
      "3100000 alignment records processed.\n",
      "3200000 alignment records processed.\n",
      "3300000 alignment records processed.\n",
      "3400000 alignment records processed.\n",
      "3500000 alignment records processed.\n",
      "3600000 alignment records processed.\n",
      "3700000 alignment records processed.\n",
      "3800000 alignment records processed.\n",
      "3900000 alignment records processed.\n",
      "4000000 alignment records processed.\n",
      "4100000 alignment records processed.\n",
      "4200000 alignment records processed.\n",
      "4300000 alignment records processed.\n",
      "4400000 alignment records processed.\n",
      "4500000 alignment records processed.\n",
      "4600000 alignment records processed.\n",
      "4700000 alignment records processed.\n",
      "4800000 alignment records processed.\n",
      "4900000 alignment records processed.\n",
      "5000000 alignment records processed.\n",
      "5100000 alignment records processed.\n",
      "5200000 alignment records processed.\n",
      "5300000 alignment records processed.\n",
      "5400000 alignment records processed.\n",
      "5500000 alignment records processed.\n",
      "5600000 alignment records processed.\n",
      "5700000 alignment records processed.\n",
      "5800000 alignment records processed.\n",
      "5900000 alignment records processed.\n",
      "6000000 alignment records processed.\n",
      "6100000 alignment records processed.\n",
      "6200000 alignment records processed.\n",
      "6300000 alignment records processed.\n",
      "6400000 alignment records processed.\n",
      "6500000 alignment records processed.\n",
      "6600000 alignment records processed.\n",
      "6700000 alignment records processed.\n",
      "6800000 alignment records processed.\n",
      "6900000 alignment records processed.\n",
      "7000000 alignment records processed.\n",
      "7100000 alignment records processed.\n",
      "7200000 alignment records processed.\n",
      "7300000 alignment records processed.\n",
      "7400000 alignment records processed.\n",
      "7500000 alignment records processed.\n",
      "7600000 alignment records processed.\n",
      "7700000 alignment records processed.\n",
      "7800000 alignment records processed.\n",
      "7900000 alignment records processed.\n",
      "8000000 alignment records processed.\n",
      "8100000 alignment records processed.\n",
      "8200000 alignment records processed.\n",
      "8300000 alignment records processed.\n",
      "8400000 alignment records processed.\n",
      "8500000 alignment records processed.\n",
      "8600000 alignment records processed.\n",
      "8700000 alignment records processed.\n",
      "8800000 alignment records processed.\n",
      "8900000 alignment records processed.\n",
      "9000000 alignment records processed.\n",
      "9100000 alignment records processed.\n",
      "9200000 alignment records processed.\n",
      "9300000 alignment records processed.\n",
      "9400000 alignment records processed.\n",
      "9500000 alignment records processed.\n",
      "9600000 alignment records processed.\n",
      "9700000 alignment records processed.\n",
      "9800000 alignment records processed.\n",
      "9900000 alignment records processed.\n",
      "10000000 alignment records processed.\n",
      "10100000 alignment records processed.\n",
      "10200000 alignment records processed.\n",
      "10300000 alignment records processed.\n",
      "10400000 alignment records processed.\n",
      "10500000 alignment records processed.\n",
      "10600000 alignment records processed.\n",
      "10700000 alignment records processed.\n",
      "10800000 alignment records processed.\n",
      "10900000 alignment records processed.\n",
      "11000000 alignment records processed.\n",
      "11100000 alignment records processed.\n",
      "11200000 alignment records processed.\n",
      "11300000 alignment records processed.\n",
      "11400000 alignment records processed.\n",
      "11500000 alignment records processed.\n",
      "11600000 alignment records processed.\n",
      "11700000 alignment records processed.\n",
      "11800000 alignment records processed.\n",
      "11900000 alignment records processed.\n",
      "12000000 alignment records processed.\n",
      "12100000 alignment records processed.\n",
      "12200000 alignment records processed.\n",
      "12300000 alignment records processed.\n",
      "12400000 alignment records processed.\n",
      "12500000 alignment records processed.\n",
      "12600000 alignment records processed.\n",
      "12700000 alignment records processed.\n",
      "12800000 alignment records processed.\n",
      "12900000 alignment records processed.\n",
      "13000000 alignment records processed.\n",
      "13100000 alignment records processed.\n",
      "13200000 alignment records processed.\n",
      "13300000 alignment records processed.\n",
      "13400000 alignment records processed.\n",
      "13500000 alignment records processed.\n",
      "13600000 alignment records processed.\n",
      "13700000 alignment records processed.\n",
      "13800000 alignment records processed.\n",
      "13900000 alignment records processed.\n",
      "14000000 alignment records processed.\n",
      "14100000 alignment records processed.\n",
      "14200000 alignment records processed.\n",
      "14300000 alignment records processed.\n",
      "14400000 alignment records processed.\n",
      "14500000 alignment records processed.\n",
      "14600000 alignment records processed.\n",
      "14700000 alignment records processed.\n",
      "14800000 alignment records processed.\n",
      "14900000 alignment records processed.\n",
      "15000000 alignment records processed.\n",
      "15100000 alignment records processed.\n",
      "15200000 alignment records processed.\n",
      "15300000 alignment records processed.\n",
      "15400000 alignment records processed.\n",
      "15500000 alignment records processed.\n",
      "15600000 alignment records processed.\n",
      "15700000 alignment records processed.\n",
      "15800000 alignment records processed.\n",
      "15900000 alignment records processed.\n",
      "16000000 alignment records processed.\n",
      "16100000 alignment records processed.\n",
      "16200000 alignment records processed.\n",
      "16300000 alignment records processed.\n",
      "16400000 alignment records processed.\n",
      "16500000 alignment records processed.\n",
      "16600000 alignment records processed.\n",
      "16700000 alignment records processed.\n",
      "16800000 alignment records processed.\n",
      "16900000 alignment records processed.\n",
      "17000000 alignment records processed.\n",
      "17100000 alignment records processed.\n",
      "17200000 alignment records processed.\n",
      "17300000 alignment records processed.\n",
      "17400000 alignment records processed.\n",
      "17500000 alignment records processed.\n",
      "17600000 alignment records processed.\n",
      "17700000 alignment records processed.\n",
      "17800000 alignment records processed.\n",
      "17900000 alignment records processed.\n",
      "18000000 alignment records processed.\n",
      "18100000 alignment records processed.\n",
      "18200000 alignment records processed.\n",
      "18300000 alignment records processed.\n",
      "18400000 alignment records processed.\n",
      "18500000 alignment records processed.\n",
      "18600000 alignment records processed.\n",
      "18700000 alignment records processed.\n",
      "18800000 alignment records processed.\n",
      "18900000 alignment records processed.\n",
      "19000000 alignment records processed.\n",
      "19100000 alignment records processed.\n",
      "19200000 alignment records processed.\n",
      "19300000 alignment records processed.\n",
      "19400000 alignment records processed.\n",
      "19500000 alignment records processed.\n",
      "19600000 alignment records processed.\n",
      "19700000 alignment records processed.\n",
      "19800000 alignment records processed.\n",
      "19900000 alignment records processed.\n",
      "20000000 alignment records processed.\n",
      "20100000 alignment records processed.\n",
      "20200000 alignment records processed.\n",
      "20300000 alignment records processed.\n",
      "20400000 alignment records processed.\n",
      "20500000 alignment records processed.\n",
      "20600000 alignment records processed.\n",
      "20700000 alignment records processed.\n",
      "20800000 alignment records processed.\n",
      "20900000 alignment records processed.\n",
      "21000000 alignment records processed.\n",
      "21100000 alignment records processed.\n",
      "21200000 alignment records processed.\n",
      "21300000 alignment records processed.\n",
      "21400000 alignment records processed.\n",
      "21500000 alignment records processed.\n",
      "21600000 alignment records processed.\n",
      "21700000 alignment records processed.\n",
      "21800000 alignment records processed.\n",
      "21900000 alignment records processed.\n",
      "22000000 alignment records processed.\n",
      "22100000 alignment records processed.\n",
      "22200000 alignment records processed.\n",
      "22300000 alignment records processed.\n",
      "22400000 alignment records processed.\n",
      "22500000 alignment records processed.\n",
      "22600000 alignment records processed.\n",
      "22700000 alignment records processed.\n",
      "22800000 alignment records processed.\n",
      "22900000 alignment records processed.\n",
      "23000000 alignment records processed.\n",
      "23100000 alignment records processed.\n",
      "23200000 alignment records processed.\n",
      "23300000 alignment records processed.\n",
      "23400000 alignment records processed.\n",
      "23500000 alignment records processed.\n",
      "23600000 alignment records processed.\n",
      "23700000 alignment records processed.\n",
      "23800000 alignment records processed.\n",
      "23900000 alignment records processed.\n",
      "24000000 alignment records processed.\n",
      "24100000 alignment records processed.\n",
      "24200000 alignment records processed.\n",
      "24300000 alignment records processed.\n",
      "24400000 alignment records processed.\n",
      "24500000 alignment records processed.\n",
      "24600000 alignment records processed.\n",
      "24700000 alignment records processed.\n",
      "24800000 alignment records processed.\n",
      "24900000 alignment records processed.\n",
      "25000000 alignment records processed.\n",
      "25100000 alignment records processed.\n",
      "25200000 alignment records processed.\n",
      "25300000 alignment records processed.\n",
      "25400000 alignment records processed.\n",
      "25500000 alignment records processed.\n",
      "25600000 alignment records processed.\n",
      "25700000 alignment records processed.\n",
      "25800000 alignment records processed.\n",
      "25900000 alignment records processed.\n",
      "26000000 alignment records processed.\n",
      "26100000 alignment records processed.\n",
      "26200000 alignment records processed.\n",
      "26300000 alignment records processed.\n",
      "26400000 alignment records processed.\n",
      "26500000 alignment records processed.\n",
      "26600000 alignment records processed.\n",
      "26700000 alignment records processed.\n",
      "26800000 alignment records processed.\n",
      "26900000 alignment records processed.\n",
      "27000000 alignment records processed.\n",
      "27100000 alignment records processed.\n",
      "27200000 alignment records processed.\n",
      "27300000 alignment records processed.\n",
      "27400000 alignment records processed.\n",
      "27500000 alignment records processed.\n",
      "27600000 alignment records processed.\n",
      "27700000 alignment records processed.\n",
      "27800000 alignment records processed.\n",
      "27900000 alignment records processed.\n",
      "28000000 alignment records processed.\n",
      "28100000 alignment records processed.\n",
      "28200000 alignment records processed.\n",
      "28300000 alignment records processed.\n",
      "28400000 alignment records processed.\n",
      "28500000 alignment records processed.\n",
      "28600000 alignment records processed.\n",
      "28700000 alignment records processed.\n",
      "28800000 alignment records processed.\n",
      "28900000 alignment records processed.\n",
      "29000000 alignment records processed.\n",
      "29100000 alignment records processed.\n",
      "29200000 alignment records processed.\n",
      "29300000 alignment records processed.\n",
      "29400000 alignment records processed.\n",
      "29500000 alignment records processed.\n",
      "29600000 alignment records processed.\n",
      "29700000 alignment records processed.\n",
      "29800000 alignment records processed.\n",
      "29900000 alignment records processed.\n",
      "30000000 alignment records processed.\n",
      "30100000 alignment records processed.\n",
      "30200000 alignment records processed.\n",
      "30300000 alignment records processed.\n",
      "30400000 alignment records processed.\n",
      "30500000 alignment records processed.\n",
      "30600000 alignment records processed.\n",
      "30700000 alignment records processed.\n",
      "30800000 alignment records processed.\n",
      "30900000 alignment records processed.\n",
      "31000000 alignment records processed.\n",
      "31100000 alignment records processed.\n",
      "31200000 alignment records processed.\n",
      "31300000 alignment records processed.\n",
      "31400000 alignment records processed.\n",
      "31500000 alignment records processed.\n",
      "31600000 alignment records processed.\n",
      "31700000 alignment records processed.\n",
      "31800000 alignment records processed.\n",
      "31900000 alignment records processed.\n",
      "32000000 alignment records processed.\n",
      "32100000 alignment records processed.\n",
      "32200000 alignment records processed.\n",
      "32300000 alignment records processed.\n",
      "32400000 alignment records processed.\n",
      "32500000 alignment records processed.\n",
      "32600000 alignment records processed.\n",
      "32700000 alignment records processed.\n",
      "32800000 alignment records processed.\n",
      "32900000 alignment records processed.\n",
      "33000000 alignment records processed.\n",
      "33100000 alignment records processed.\n",
      "33200000 alignment records processed.\n",
      "33300000 alignment records processed.\n",
      "33400000 alignment records processed.\n",
      "33500000 alignment records processed.\n",
      "33600000 alignment records processed.\n",
      "33700000 alignment records processed.\n",
      "33800000 alignment records processed.\n",
      "33900000 alignment records processed.\n",
      "34000000 alignment records processed.\n",
      "34100000 alignment records processed.\n",
      "34200000 alignment records processed.\n",
      "34300000 alignment records processed.\n",
      "34400000 alignment records processed.\n",
      "34500000 alignment records processed.\n",
      "34600000 alignment records processed.\n",
      "34700000 alignment records processed.\n",
      "34800000 alignment records processed.\n",
      "34900000 alignment records processed.\n",
      "35000000 alignment records processed.\n",
      "35100000 alignment records processed.\n",
      "35200000 alignment records processed.\n",
      "35300000 alignment records processed.\n",
      "35400000 alignment records processed.\n",
      "35500000 alignment records processed.\n",
      "35600000 alignment records processed.\n",
      "35700000 alignment records processed.\n",
      "35800000 alignment records processed.\n",
      "35900000 alignment records processed.\n",
      "36000000 alignment records processed.\n",
      "36100000 alignment records processed.\n",
      "36200000 alignment records processed.\n",
      "36300000 alignment records processed.\n",
      "36400000 alignment records processed.\n",
      "36500000 alignment records processed.\n",
      "36600000 alignment records processed.\n",
      "36700000 alignment records processed.\n",
      "36800000 alignment records processed.\n",
      "36900000 alignment records processed.\n",
      "37000000 alignment records processed.\n",
      "37100000 alignment records processed.\n",
      "37200000 alignment records processed.\n",
      "37300000 alignment records processed.\n",
      "37400000 alignment records processed.\n",
      "37500000 alignment records processed.\n",
      "37600000 alignment records processed.\n",
      "37700000 alignment records processed.\n",
      "37800000 alignment records processed.\n",
      "37900000 alignment records processed.\n",
      "38000000 alignment records processed.\n",
      "38100000 alignment records processed.\n",
      "38200000 alignment records processed.\n",
      "38300000 alignment records processed.\n",
      "38400000 alignment records processed.\n",
      "38500000 alignment records processed.\n",
      "38600000 alignment records processed.\n",
      "38700000 alignment records processed.\n",
      "38800000 alignment records processed.\n",
      "38873373 alignment records processed.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/ssobti/projects/heterogeneity_brian/output_data/CRISPRi_CRISPRa_cell_culture_screen/CRISPRi_STAR_guide_reads_alignment\n",
    "nohup htseq-count -n 16 -f bam -s yes -i gene_id SL399_R2_all_reads_no_trimAligned.sortedByCoord.out.bam /home/genomes/hg38/star_hg38/Homo_sapiens.GRCh38.99.gtf > SL399_R2_all_reads_no_trim_HTSeq_count.txt &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T17:00:16.236047Z",
     "iopub.status.busy": "2023-08-29T17:00:16.235713Z",
     "iopub.status.idle": "2023-08-29T17:00:17.138993Z",
     "shell.execute_reply": "2023-08-29T17:00:17.138156Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60680</th>\n",
       "      <td>__alignment_not_unique</td>\n",
       "      <td>4952338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60676</th>\n",
       "      <td>__no_feature</td>\n",
       "      <td>2703444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60677</th>\n",
       "      <td>__ambiguous</td>\n",
       "      <td>478201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21453</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>73514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1315</th>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>55308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28657</th>\n",
       "      <td>ENSG00000229325</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28658</th>\n",
       "      <td>ENSG00000229326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28659</th>\n",
       "      <td>ENSG00000229327</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28660</th>\n",
       "      <td>ENSG00000229330</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30340</th>\n",
       "      <td>ENSG00000231794</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60681 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gene    count\n",
       "60680  __alignment_not_unique  4952338\n",
       "60676            __no_feature  2703444\n",
       "60677             __ambiguous   478201\n",
       "21453         ENSG00000211459    73514\n",
       "1315          ENSG00000075624    55308\n",
       "...                       ...      ...\n",
       "28657         ENSG00000229325        0\n",
       "28658         ENSG00000229326        0\n",
       "28659         ENSG00000229327        0\n",
       "28660         ENSG00000229330        0\n",
       "30340         ENSG00000231794        0\n",
       "\n",
       "[60681 rows x 2 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## load in bed file gene counts from reads\n",
    "import pandas as pd\n",
    "import mygene as mg\n",
    "\n",
    "path = '/home/ssobti/projects/heterogeneity_brian/output_data/CRISPRi_CRISPRa_cell_culture_screen/CRISPRi_STAR_guide_reads_alignment/'\n",
    "read_counts = pd.read_csv(path + 'SL399_R2_all_reads_no_trim_HTSeq_count.txt', sep='\\t', names = ['gene', 'count'])\n",
    "read_counts.sort_values(by='count', ascending=False, inplace=True)\n",
    "read_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T17:00:17.144949Z",
     "iopub.status.busy": "2023-08-29T17:00:17.144586Z",
     "iopub.status.idle": "2023-08-29T17:00:17.167415Z",
     "shell.execute_reply": "2023-08-29T17:00:17.166439Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21453</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>73514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1315</td>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>55308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21417</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>42099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6598</td>\n",
       "      <td>ENSG00000132507</td>\n",
       "      <td>30667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7539</td>\n",
       "      <td>ENSG00000137818</td>\n",
       "      <td>19020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60673</th>\n",
       "      <td>28657</td>\n",
       "      <td>ENSG00000229325</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60674</th>\n",
       "      <td>28658</td>\n",
       "      <td>ENSG00000229326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60675</th>\n",
       "      <td>28659</td>\n",
       "      <td>ENSG00000229327</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60676</th>\n",
       "      <td>28660</td>\n",
       "      <td>ENSG00000229330</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60677</th>\n",
       "      <td>30340</td>\n",
       "      <td>ENSG00000231794</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60678 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index             gene  count\n",
       "0      21453  ENSG00000211459  73514\n",
       "1       1315  ENSG00000075624  55308\n",
       "2      21417  ENSG00000210082  42099\n",
       "3       6598  ENSG00000132507  30667\n",
       "4       7539  ENSG00000137818  19020\n",
       "...      ...              ...    ...\n",
       "60673  28657  ENSG00000229325      0\n",
       "60674  28658  ENSG00000229326      0\n",
       "60675  28659  ENSG00000229327      0\n",
       "60676  28660  ENSG00000229330      0\n",
       "60677  30340  ENSG00000231794      0\n",
       "\n",
       "[60678 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove_rows = ['__alignment_not_unique', '__no_feature', '__ambiguous']\n",
    "read_counts = read_counts[~read_counts['gene'].isin(remove_rows)].reset_index()\n",
    "read_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T17:00:17.171761Z",
     "iopub.status.busy": "2023-08-29T17:00:17.171452Z",
     "iopub.status.idle": "2023-08-29T17:00:17.254865Z",
     "shell.execute_reply": "2023-08-29T17:00:17.253938Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "querying 2001-3000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 7001-8000...done.\n",
      "querying 8001-9000...done.\n",
      "querying 9001-10000...done.\n",
      "querying 10001-11000...done.\n",
      "querying 11001-12000...done.\n",
      "querying 12001-13000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 14001-15000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 19001-20000...done.\n",
      "querying 20001-21000...done.\n",
      "querying 21001-22000...done.\n",
      "querying 22001-23000...done.\n",
      "querying 23001-24000...done.\n",
      "querying 24001-25000...done.\n",
      "querying 25001-26000...done.\n",
      "querying 26001-27000...done.\n",
      "querying 27001-28000...done.\n",
      "querying 28001-29000...done.\n",
      "querying 29001-30000...done.\n",
      "querying 30001-31000...done.\n",
      "querying 31001-32000...done.\n",
      "querying 32001-33000...done.\n",
      "querying 33001-34000...done.\n",
      "querying 34001-35000...done.\n",
      "querying 35001-36000...done.\n",
      "querying 36001-37000...done.\n",
      "querying 37001-38000...done.\n",
      "querying 38001-39000...done.\n",
      "querying 39001-40000...done.\n",
      "querying 40001-41000...done.\n",
      "querying 41001-42000...done.\n",
      "querying 42001-43000...done.\n",
      "querying 43001-44000...done.\n",
      "querying 44001-45000...done.\n",
      "querying 45001-46000...done.\n",
      "querying 46001-47000...done.\n",
      "querying 47001-48000...done.\n",
      "querying 48001-49000...done.\n",
      "querying 49001-50000...done.\n",
      "querying 50001-51000...done.\n",
      "querying 51001-52000...done.\n",
      "querying 52001-53000...done.\n",
      "querying 53001-54000...done.\n",
      "querying 54001-55000...done.\n",
      "querying 55001-56000...done.\n",
      "querying 56001-57000...done.\n",
      "querying 57001-58000...done.\n",
      "querying 58001-59000...done.\n",
      "querying 59001-60000...done.\n",
      "querying 60001-60678...done.\n",
      "Finished.\n",
      "267 input query terms found no hit:\n",
      "\t['ENSG00000130723', 'ENSG00000254615', 'ENSG00000269926', 'ENSG00000112096', 'ENSG00000239665', 'ENS\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"
     ]
    }
   ],
   "source": [
    "## convert ENSEMBL IDs to gene symbols\n",
    "import mygene as mg\n",
    "mg = mg.MyGeneInfo()\n",
    "gene_symbols = mg.querymany(read_counts['gene'].tolist(), scopes='ensembl.gene', fields = ['symbol'], species = 'human')\n",
    "gene_symbols = [gene['symbol'] if 'symbol' in gene.keys() else 'NA' for gene in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T17:00:17.260115Z",
     "iopub.status.busy": "2023-08-29T17:00:17.259664Z",
     "iopub.status.idle": "2023-08-29T17:00:17.310814Z",
     "shell.execute_reply": "2023-08-29T17:00:17.310150Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21453</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>73514</td>\n",
       "      <td>RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1315</td>\n",
       "      <td>ENSG00000075624</td>\n",
       "      <td>55308</td>\n",
       "      <td>ACTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21417</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>42099</td>\n",
       "      <td>RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6598</td>\n",
       "      <td>ENSG00000132507</td>\n",
       "      <td>30667</td>\n",
       "      <td>EIF5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7539</td>\n",
       "      <td>ENSG00000137818</td>\n",
       "      <td>19020</td>\n",
       "      <td>RPLP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17819</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>18344</td>\n",
       "      <td>CYTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>17572</td>\n",
       "      <td>ENSG00000197956</td>\n",
       "      <td>16549</td>\n",
       "      <td>S100A6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>14990</td>\n",
       "      <td>ENSG00000181163</td>\n",
       "      <td>15670</td>\n",
       "      <td>NPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>16805</td>\n",
       "      <td>ENSG00000188846</td>\n",
       "      <td>13474</td>\n",
       "      <td>RPL14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3492</td>\n",
       "      <td>ENSG00000108107</td>\n",
       "      <td>11799</td>\n",
       "      <td>RPL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>452</td>\n",
       "      <td>ENSG00000026025</td>\n",
       "      <td>11737</td>\n",
       "      <td>VIM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>8290</td>\n",
       "      <td>ENSG00000142937</td>\n",
       "      <td>11503</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13739</td>\n",
       "      <td>ENSG00000174748</td>\n",
       "      <td>10849</td>\n",
       "      <td>RPL15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17811</td>\n",
       "      <td>ENSG00000198712</td>\n",
       "      <td>10511</td>\n",
       "      <td>COX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9263</td>\n",
       "      <td>ENSG00000149925</td>\n",
       "      <td>10108</td>\n",
       "      <td>ALDOA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>53721</td>\n",
       "      <td>ENSG00000276168</td>\n",
       "      <td>8137</td>\n",
       "      <td>RN7SL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17805</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>7579</td>\n",
       "      <td>ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17941</td>\n",
       "      <td>ENSG00000198938</td>\n",
       "      <td>6883</td>\n",
       "      <td>COX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1718</td>\n",
       "      <td>ENSG00000087086</td>\n",
       "      <td>6742</td>\n",
       "      <td>FTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>10443</td>\n",
       "      <td>ENSG00000160213</td>\n",
       "      <td>6480</td>\n",
       "      <td>CSTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3543</td>\n",
       "      <td>ENSG00000108518</td>\n",
       "      <td>6315</td>\n",
       "      <td>PFN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>16238</td>\n",
       "      <td>ENSG00000186468</td>\n",
       "      <td>6129</td>\n",
       "      <td>RPS23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5805</td>\n",
       "      <td>ENSG00000126261</td>\n",
       "      <td>5868</td>\n",
       "      <td>UBA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>10566</td>\n",
       "      <td>ENSG00000161016</td>\n",
       "      <td>5667</td>\n",
       "      <td>RPL8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>5611</td>\n",
       "      <td>ENSG00000125148</td>\n",
       "      <td>5573</td>\n",
       "      <td>MT2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>17848</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>5502</td>\n",
       "      <td>ND5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3913</td>\n",
       "      <td>ENSG00000111640</td>\n",
       "      <td>5397</td>\n",
       "      <td>GAPDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>19512</td>\n",
       "      <td>ENSG00000202515</td>\n",
       "      <td>5285</td>\n",
       "      <td>VTRNA1-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>13376</td>\n",
       "      <td>ENSG00000172809</td>\n",
       "      <td>5276</td>\n",
       "      <td>RPL38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>17909</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>5213</td>\n",
       "      <td>ND4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index             gene  count    symbol\n",
       "0   21453  ENSG00000211459  73514      RNR1\n",
       "1    1315  ENSG00000075624  55308      ACTB\n",
       "2   21417  ENSG00000210082  42099      RNR2\n",
       "3    6598  ENSG00000132507  30667     EIF5A\n",
       "4    7539  ENSG00000137818  19020     RPLP1\n",
       "5   17819  ENSG00000198727  18344      CYTB\n",
       "6   17572  ENSG00000197956  16549    S100A6\n",
       "7   14990  ENSG00000181163  15670      NPM1\n",
       "8   16805  ENSG00000188846  13474     RPL14\n",
       "9    3492  ENSG00000108107  11799     RPL28\n",
       "10    452  ENSG00000026025  11737       VIM\n",
       "11   8290  ENSG00000142937  11503      RPS8\n",
       "12  13739  ENSG00000174748  10849     RPL15\n",
       "13  17811  ENSG00000198712  10511      COX2\n",
       "14   9263  ENSG00000149925  10108     ALDOA\n",
       "15  53721  ENSG00000276168   8137    RN7SL1\n",
       "16  17805  ENSG00000198695   7579       ND6\n",
       "17  17941  ENSG00000198938   6883      COX3\n",
       "18   1718  ENSG00000087086   6742       FTL\n",
       "19  10443  ENSG00000160213   6480      CSTB\n",
       "20   3543  ENSG00000108518   6315      PFN1\n",
       "21  16238  ENSG00000186468   6129     RPS23\n",
       "22   5805  ENSG00000126261   5868      UBA2\n",
       "23  10566  ENSG00000161016   5667      RPL8\n",
       "24   5611  ENSG00000125148   5573      MT2A\n",
       "25  17848  ENSG00000198786   5502       ND5\n",
       "26   3913  ENSG00000111640   5397     GAPDH\n",
       "27  19512  ENSG00000202515   5285  VTRNA1-3\n",
       "28  13376  ENSG00000172809   5276     RPL38\n",
       "29  17909  ENSG00000198886   5213       ND4"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## adding gene symbols\n",
    "read_counts['symbol'] = gene_symbols\n",
    "read_counts[0:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "seq_trblsht",
   "language": "python",
   "name": "seq_trblsht"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
